lua/sql/table.lua

```diff
@@ -0,0 +1,251 @@
+local u = require'sql.utils'
+local t = {}
+t.__index = t
+
+function t:__run(func)
+  if self.db.closed then
+    return self.db:with_open(function()
+      return func() -- shoud pass tbl name?
+    end)
+  else
+   return func()
+  end
+end
+
+local get_key = function(query)
+  local items = {}
+  for k, v in u.opairs(query) do
+    table.insert(items, string.format("%s=%s", k, table.concat((function()
+      if not u.is_list(v) then
+        local tmp = {}
+        for _k,_v in u.opairs(v) do
+          if type(_v) == "table" then
+            table.insert(tmp, string.format("%s=%s", _k, table.concat(_v, ",")))
+            else
+          table.insert(tmp, string.format("%s=%s", _k, _v))
+          end
+        end
+        return tmp
+      else
+        return v
+      end
+    end)(), "")))
+  end
+
+  return table.concat(items, ",")
+end
+
+
+function t:__clear_cache(succ, change)
+  -- TODO: maybe do something smart here to update cache :D
+  if succ then
+    self.cache = {}
+    self.db.modified = false
+  end
+end
+
+function t:__fill_cache(query, res)
+  self.cache[get_key(query)] = res
+end
+
+function t:__get_from_cache(query)
+  if vim.loop.fs_stat(self.db.uri).mtime.sec ~= self.mtime or self.db.modified then
```

Comment:
Do the `self.db.modified` first so you are not always doing the filesystem access

Reply:
Great idea conni